device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubSendTask.java

```diff
@@ -21,7 +23,7 @@
     // as this SDK would otherwise periodically spawn new threads of this type that would do nothing. It is the IotHubTransport
     // layer's responsibility to notify this thread when a message is queued to be sent or when a callback is queued to be executed
     // so that this thread can handle it.
-    private final Object sendThreadLock;
+    private final Semaphore sendThreadSemaphore;
```

Comment:
Changing this to a semaphore to avoid any potential out-of-order problems where the lock is notified but no thread is waiting for it since this could cause deadlocks. With semaphores, the release can be called prior to anyone trying to acquire it, so the order of operations no longer matters

Reply:
This isn't as big of a deal in send/receive threads since typically users send/receive more than one message per device client, but for the new reconnect thread, this is very important since the disconnection event is much more rare.